### DISCUSSION TITLE
Is there a way to define default "_format" for internal requests to \Symfony\Component\HttpKernel\Controller\ErrorController ?

### DISCUSSION THREAD

**Discussion by flavacaster** [symfony/symfony#58801] [Q&A] [answered]
Hi everyone. You can define the default format (e.g. json) for your routes like this 

<img width="561" alt="PhpStorm 2024-11-07 16 03 46" src="https://github.com/user-attachments/assets/e9c449ca-2ee7-4b8f-b85e-6a6c0253b02a">

this way the response will contain JSON in case any exception is thrown during the execution. This works only in case the route was successfully matched. 

But if the route could not be matched (e.g. an HTTP method that is not supported by the route was used) the default format is not set, therefore \Symfony\Component\HttpKernel\Controller\ErrorController (more specifically \Symfony\Component\ErrorHandler\ErrorRenderer\SerializerErrorRenderer) falls back to html format. 

As a workaround I've created this EventListener 

<img width="782" alt="image" src="https://github.com/user-attachments/assets/70ce1345-7acb-43b3-966e-4711e8d84f87">

I feel like there should be an option to set the default format for ErrorController route "out of the box" given that we can easily set the format for our custom defined routes

**Comment by MatTheCat** [comment, marked as answer]
Hello,

by default Symfony will honor the `Accept` request header so you have the option to set it to `application/json`.

Else you can listen to the `kernel.exception` event to set the request format:

```php
<?php

namespace App;

use Symfony\Component\EventDispatcher\Attribute\AsEventListener;
use Symfony\Component\HttpKernel\Event\ExceptionEvent;

#[AsEventListener('kernel.exception')]
class ChangeRequestFormatOnException
{
    public function __invoke(ExceptionEvent $event): void
    {
        $event->getRequest()->setRequestFormat('json');
    }
}
```
